drivers/char/optimsoc-noc.c

```diff
@@ -0,0 +1,436 @@
+		if (adapters[i].data_received)
+			wake_up(&adapters[i].wqueue);
+	}
+
+	return (IRQ_HANDLED);
+}
+
+/*
+ * Opens the NoC device.
+ */
+static int device_open(struct inode *inode, struct file *file)
+{
+	unsigned minor;
+
+	minor = iminor(inode);
+
+	printk(KERN_INFO "%s: open device %d", OPTIMSOC_NA_NAME, minor);
+
+	/* Device already in use. */
+	if (adapters[minor].nopen > 0)
+		return (-EBUSY);
+
+	/* Grab resources. */
+    adapters[minor].buffer = kmalloc(OPTIMSOC_NA_BUFFER_SIZE, GFP_KERNEL);
+	if (adapters[minor].buffer == NULL)
+		return (-ENOMEM);
+
+	adapters[minor].nopen++;
+
+	enable(minor);
+
+	try_module_get(THIS_MODULE);
+
+	return (SUCCESS);
+}
+
+/*
+ * Closes the NoC device.
+ */
+static int device_release(struct inode *inode, struct file *file)
+{
+	unsigned minor;
+
+	minor = iminor(inode);
+
+	printk(KERN_INFO "%s: close device %d", OPTIMSOC_NA_NAME, minor);
+
+	/* Release resources. */
+	adapters[minor].nopen--;
+	kfree(adapters[minor].buffer);
+	adapters[minor].buffer = NULL;
+
+	module_put(THIS_MODULE);
+
+	return (SUCCESS);
+}
+
+/*
+ * Reads bytes from the NoC device.
+ */
+static ssize_t device_read(struct file *filp, char *buffer, size_t length, loff_t * offset)
+{
+	size_t i;
+	unsigned minor;
+
+	((void) offset);
+
+	minor = iminor(filp->f_inode);
+	if (minor >= nr_endpoints)
+		return (-EINVAL);
+
+	printk(KERN_INFO "%s: read from device %d", OPTIMSOC_NA_NAME, minor);
+
+	/* Read bytes. */
+	for (i = 0; i < length; /* noop*/)
+	{
+		size_t n;
+		uint32_t word;
+		size_t count;
+
+		/* Wait for data, */
+		if (adapters[minor].head == adapters[minor].tail)
+		{
+			wait_event_interruptible(adapters[minor].wqueue, adapters[minor].data_received);
+			adapters[minor].data_received = !adapters[minor].data_received;
+		}
+
+		word = adapters[minor].buffer[adapters[minor].head];
+		adapters[minor].head = (adapters[minor].head + 1)%OPTIMSOC_NA_BUFFER_SIZE;
+
+		n = ((i + 4) <= length) ? 4 : (length  - i);
+
+		count = copy_to_user(&buffer[i], &word, n);
+
+		printk(KERN_INFO "receive: %x", word);
+
+		if (count != 0)
+			return (i + (n - count));
+
+		i += 4;
+	}
+
+	return (length);
+}
+
+/*
+ * Writes bytes from the NoC device.
+ */
+static ssize_t device_write(struct file *filp, const char *buff, size_t len, loff_t * off)
+{
+	int i;
+	unsigned minor;
+
+	((void) off);
+
+	/* Get minor device. */
+	minor = iminor(filp->f_inode);
+	if (minor >= nr_endpoints)
+		return (-EINVAL);
+
+	printk(KERN_INFO "%s: write to device %d", OPTIMSOC_NA_NAME, minor);
+
+	/* Copy data to temporary buffer. */
+	for (i = 0; i < len; /* noop. */)
+	{
+		size_t n;
+		size_t count;
+		uint32_t word;
+
+		n = ((i + 4) <= len) ? 4 : (len  - i);
+
+		count = copy_from_user(&word, &buff[i], n);
+		send(minor, word);
+
+		if (count != 0)
+			return (i + (n - count));
+
+		i += 4;
+	}
+
+	return (len);
+}
+
+static struct file_operations fops = {
+	.read = device_read,
+	.write = device_write,
+	.open = device_open,
+	.release = device_release
```

Comment:
If you name those optimsoc_noc_read, etc. it'll be easier to find in traces. (Unless it's common practice to do otherwise in other drivers, of course.)